File: Algorithm/4/intersect.cpp
#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        // unordered_map<int, int> record;
        map<int, int> record;
        for(int i = 0; i < nums1.size(); i++)
            record[nums1[i]]++;
            // if(record.find(nums1[i]) == record.end())
            //     record.insert(make_pair(nums1[i], 1));
            // else
            //     record[nums1[i]]++
        vector<int> resultVector;
        for(int i = 0; i < nums2.size(); i++)
            if(record[nums2[i]] > 0){
                resultVector.push_back(nums2[i]);
                record[nums2[i]]--;
            }
            // if(record.find(nums2[i]) != record.end() && record[nums2[i]] > 0){
            //     resultVector.push_back(nums2[i]);
            //     record[nums2[i]]--;
            //     if(record[nums2[i]] == 0)
            //         record.erase(nums2[i]);
            // }
        return resultVector;
    }
};

File: Algorithm/4/intersect.py
def intersect(nums1, nums2):
    """
    :type nums1: List[int]
    :type nums2: List[int]
    :rtype: List[int]
    """
    res = []
    for i in nums1:
        if i in nums2:
            res.append(i)
            nums2.remove(i)
    return res

def intersect(nums1, nums2):
    """
    :type nums1: List[int]
    :type nums2: List[int]
    :rtype: List[int]
    """
    res = {}
    ans = []
    for i in nums1:
        if i not in res:
            res[i] = 1
        else:
            res[i] += 1
    for i in nums2:
        if i in res and res[i] > 0:
            ans.append(i)
            res[i] -= 1
    return ans
